ui/src/logs/components/expandable_message/ExpandedContainer.tsx

```diff
@@ -0,0 +1,151 @@
+// Libraries
+import React, {Component, MouseEvent, CSSProperties} from 'react'
+
+// Components
+import {ClickOutside} from 'src/shared/components/ClickOutside'
+import FancyScrollbar from 'src/shared/components/FancyScrollbar'
+
+// Types
+import {NotificationAction} from 'src/types'
+
+// Decorators
+import {ErrorHandling} from 'src/shared/decorators/errors'
+
+interface Props {
+  notify: NotificationAction
+  onClose: () => void
+  maxWidth: number
+  maxHeight: number
+  minWidth: number
+  minLeft: number
+  width: number
+  left: number
+  top: number
+  padding: number
+  children: JSX.Element
+}
+
+interface State {
+  scrollTop: number
+}
+
+const MIN_LEFT = 120
+
+@ErrorHandling
+export class ExpandedContainer extends Component<Props, State> {
+  private initialVisibility: boolean
+
+  constructor(props: Props) {
+    super(props)
+
+    this.initialVisibility = this.isVisible
+    this.state = {scrollTop: 0}
+  }
+
+  public componentDidMount() {
+    if (!this.initialVisibility) {
+      this.props.onClose()
+    }
+  }
+
+  public render() {
+    return (
+      <ClickOutside onClickOutside={this.handleClickOutside}>
+        {this.renderMessage({
+          top: this.top,
+          left: this.maxLeft,
+          width: this.visibleWidth,
+          padding: this.props.padding,
+          maxHeight: this.maxHeight,
+        })}
+      </ClickOutside>
+    )
+  }
+
+  private renderMessage(style: CSSProperties): JSX.Element {
+    if (style.width < this.props.minWidth) {
+      return this.props.children
+    }
+
+    return (
+      <div className="expanded--message message-wrap" style={style}>
+        {this.closeExpansionButton}
+        <FancyScrollbar
+          setScrollTop={this.handleScrollbarScroll}
+          scrollTop={this.state.scrollTop}
+          autoHeight={true}
+          maxHeight={this.props.maxHeight}
+        >
+          {this.props.children}
+        </FancyScrollbar>
+      </div>
+    )
+  }
+
+  private get maxHeight(): number {
+    return this.props.maxHeight + 2 * this.props.padding
```

Comment:
what is this magic number 2? how do you feel about giving it a variable name?

Reply:
Good point! I think a getter called `verticalPadding` might work here, and I think actually subtract it here, rename `maxHeight` tp `maxScrollableHeight` and just set the container `maxHeight` to `props.maxHeight`.